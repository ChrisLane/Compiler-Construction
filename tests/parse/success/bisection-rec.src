// This is the recursive bisection algorithm. This should parse.
bisect (l, r) {
    var mid = (r - l) / 2;
    
    if (mid >= e) {
        if (f(r) * f(mid) <= 0) {
            var l = mid;
        } else {
            var r = mid;
        };
        bisect(l, r);
    } else {};

    return $mid;
}

main() {
    bisect();
}
