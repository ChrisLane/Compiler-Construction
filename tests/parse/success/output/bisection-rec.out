
bisect ( l, r ) { 
    New ( "mid" = Operator ( Divide, Operator ( Minus, Identifier "r" , Identifier "l" ) , Const 2 ) ) In { 
        If ( Operator ( Geq, Identifier "mid" , Identifier "e" ) ) { 
            If ( Operator ( Leq, Operator ( Times, Application ( Identifier "f" ( Identifier "r" ) ) , Application ( Identifier "f" ( Identifier "mid" ) ) ) , Const 0 ) ) { 
                New ( "l" = Identifier "mid" ) In { 
                    empty 
                } 
            } Else { 
                New ( "r" = Identifier "mid" ) In { 
                    empty 
                } 
            } ; 
            Application ( Identifier "bisect" ( Identifier "l" , Identifier "r" ) ) 
        } Else { 
            empty 
        } ; 
        Deref ( Identifier "mid" ) 
    } 
}

main (  ) { 
    Application ( Identifier "bisect" ( empty ) ) 
}

